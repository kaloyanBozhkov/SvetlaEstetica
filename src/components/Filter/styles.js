const styles = {
    filter: {
        flexDir: {
            xs: 'column',
            base: 'row'
        },
        justifyContent: 'space-between',
        color: 'white',
        fontSize: 'lg',
        fontWeight: 'normal',
        lineHeight: 'tight',
        marginTop: '1rem'
    },
    categorySelectWrapper: {
        gridTemplateColumns: '1fr 1fr',
        gap: '1rem',
        alignItems: 'center',
    },
    selectCategory: {
        background: 'none',
        border: 'none',
        focusBorderColor: 'none',
        borderRadius: '0px',
        borderBottom: '1px solid',
        color: 'white',
        h: '24px',
        iconSize: '18px',
        fontWeight: 'medium',
        lineHeight: '18px',
        textAlign: 'left',
        padding: '0',
    },
    priceRangeWrapper: {
        width: '20rem',
        gridTemplateColumns: '1fr minmax(12rem, 1fr)',
        gap: '1rem',
        alignItems: 'center',
    }
}

export default styles