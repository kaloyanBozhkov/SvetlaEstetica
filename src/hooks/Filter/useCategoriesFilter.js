import React, { useState } from 'react'
import Category from 'components/Filter/Category/Category'

// controller for Filter component
const useCategoriesFilter = (categories = []) => {

    // set initial filter state
    const initialState = null

    const [selectedCategory, _setSelectedCategory] = useState(initialState)

    // create reducer update fns
    const setSelectedCategory = (categoryLabel) => _setSelectedCategory(categoryLabel)
    const clearSelectedCategory = () => _setSelectedCategory(initialState)

    const categoryFilterConfig = {
        label: 'Categoria',
        component: () => <Category
            categories={categories}
            selectedCategory={selectedCategory}
            setSelectedCategory={setSelectedCategory}
        />,
        filter: 'selectedCategory'
    }

    const categoryFilter = (initialState === selectedCategory) ? null : {
        label: categoryFilterConfig.label,
        value: selectedCategory,
        onClear: clearSelectedCategory,
    }

    return {
        categoryFilterState: { selectedCategory },
        categoryFilterConfig,
        categoryFilter,
        clearSelectedCategory
    }
}

export default useCategoriesFilter