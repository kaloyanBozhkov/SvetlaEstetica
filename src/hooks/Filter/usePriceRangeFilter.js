import React, { useState } from 'react'
import PriceRange from 'components/Filter/PriceRange/PriceRange'

const applyArrayDisplayConfig = (arr, currency) => {
    const copy = [...arr]
    const formatter = function () {
        if (!this) return ''

        return this[0] + currency + ' - ' + this[1]
    }
    copy.toString = formatter
    copy.join = formatter

    return copy
}

// controller for Filter component
const usePriceRangeFilter = ({ maxPrice, minPrice, currency }) => {
    // set initial filter state
    const initialState = applyArrayDisplayConfig([minPrice, maxPrice], currency)

    const [priceRange, _setPriceRange] = useState(initialState)

    // create state update fns
    const setPriceRange = (priceRange) =>
        _setPriceRange(applyArrayDisplayConfig(priceRange, currency))

    const clearPriceRangeFilter = () => _setPriceRange(initialState)

    const priceRangeFilterConfig = {
        label: 'Prezzo',
        component: () => (
            <PriceRange
                maxPrice={maxPrice}
                minPrice={minPrice}
                priceRange={priceRange}
                setPriceRange={setPriceRange}
            />
        ),
        filter: 'priceRange',
    }

    // if filter has been set to anything but default value
    const priceRangeFilter = (initialState[0] === priceRange[0] && initialState[1] === priceRange[1]) ? null : {
        label: priceRangeFilterConfig.label,
        value: priceRange.toString(), // convert to string, so custom displayers are applied
        onClear: clearPriceRangeFilter,
    }

    return {
        priceRangeFilterState: { priceRange },
        priceRangeFilterConfig,
        priceRangeFilter,
        clearPriceRangeFilter,
    }
}

export default usePriceRangeFilter
