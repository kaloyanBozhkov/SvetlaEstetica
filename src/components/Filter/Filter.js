import React from 'react'
import { Flex, Grid, Select, Slider, SliderFilledTrack, SliderThumb, SliderTrack, Text } from '@chakra-ui/core'
import styles from './styles'


const Filter = ({ categories = [], maxPrice = 0, currentPrice = 0, selectedCategory = -1, setSelectedCategory = (f) => f, setPriceRange = (f) => f }) => {
    return (
        <Flex {...styles.filter}>
            <Grid {...styles.categorySelectWrapper}>
                <Text>Sto cercando</Text>
                <Select
                    defaultValue={selectedCategory}
                    placeholder="Tutto"
                    {...styles.selectCategory}
                    onChange={({ target: { value } }) => setSelectedCategory(value)}
                >
                    {categories.map(({ category, id }) => (
                        <option value={id} key={id}>{category}</option>
                    ))}
                </Select>
            </Grid>
            <Grid {...styles.priceRangeWrapper}>
                <Text>Prezzo</Text>
                <Slider defaultValue={currentPrice} max={maxPrice} onChange={setPriceRange}>
                    <SliderTrack>
                        <SliderFilledTrack bg="white" />
                    </SliderTrack>
                    <SliderThumb />
                </Slider>
            </Grid>
        </Flex>
    )
}

export default Filter