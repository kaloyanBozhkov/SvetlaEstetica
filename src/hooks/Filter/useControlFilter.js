import { useState } from 'react'
import useCategoriesFilter from './useCategoriesFilter'
import usePriceRangeFilter from './usePriceRangeFilter'

// controller for Filter component
const useControlFilter = ({ categories = [], maxPrice, minPrice, currency }) => {

    const {
        priceRangeFilterState,
        priceRangeFilterConfig,
        priceRangeFilter,
        clearPriceRangeFilter,
    } = usePriceRangeFilter({ maxPrice, minPrice, currency })

    const {
        categoryFilterState,
        categoryFilterConfig,
        categoryFilter,
        clearSelectedCategory
    } = useCategoriesFilter(categories)

    // clear all filters
    const clearFilters = () => {
        clearPriceRangeFilter()
        clearSelectedCategory()
    }

    // have obj with 'property':'value' of the actual state of individual filters so we can filterItems based on those
    const filters = {
        ...categoryFilterState,
        ...priceRangeFilterState
    }

    const filterConfigs = [
        categoryFilterConfig,
        priceRangeFilterConfig
    ]

    // if a filter is not null then add to chosenFilters
    const chosenFilters = []

    if (categoryFilter) {
        chosenFilters.push(categoryFilter)
    }

    if (priceRangeFilter) {
        chosenFilters.push(priceRangeFilter)
    }


    // active filter overlay -- not in reducer cuz unrelated to filter config
    const [activeFilter, setActiveFilterFn] = useState('')
    const setActiveFilter = (filterLabel) => setActiveFilterFn(activeFilter !== filterLabel ? filterLabel : '')

    return {
        activeFilter,
        setActiveFilter,
        filters,
        filterConfigs,
        chosenFilters,
        clearFilters
    }
}

export default useControlFilter