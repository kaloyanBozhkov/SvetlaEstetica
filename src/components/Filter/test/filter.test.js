import { shallow } from 'enzyme'
import React from 'react'

import Filter from '../Filter'

describe('testing Filter', () => {
    const mockProps = {
        categories: [
            {
                id: 1,
                value: 'some value 1'
            },
            {
                id: 2,
                value: 'some value 2'
            }
        ],
        setSelectedCategory: jest.fn((f) => f),
        setPriceRange: jest.fn((f) => f),
        maxPrice: 100,
        currentPrice: 10
    }
    it('should render component', () => {
        const wrapper = shallow(<Filter />)
        expect(wrapper).toMatchSnapshot()
    })

    it('should render select with options provided categories arr', () => {
        const wrapper = shallow(<Filter categories={mockProps.categories} />)
        const options = wrapper.find('option')
        expect(options.length).toEqual(mockProps.categories.length)
        expect(options.at(0).prop('value')).toEqual(mockProps.categories[0].id)
    })

    it('should run fn to change selection', () => {
        const mockSelectedValue = 2
        const wrapper = shallow(<Filter categories={mockProps.categories} setSelectedCategory={mockProps.setSelectedCategory} />)

        expect(mockProps.setSelectedCategory).toHaveBeenCalledTimes(0)

        wrapper.find('Select').prop('onChange')({ target: { value: mockSelectedValue } })

        expect(mockProps.setSelectedCategory).toHaveBeenCalled()

        // fn returns newly selected options' value
        expect(mockProps.setSelectedCategory.mock.calls[0][0]).toEqual(mockSelectedValue)
    })

    it('should set default select value if selectedCategory provided', () => {
        const mockSelectedValue = 2
        let wrapper = shallow(<Filter categories={mockProps.categories} />)
        expect(wrapper.find('Select').prop('defaultValue')).toEqual(-1)
        wrapper = shallow(<Filter categories={mockProps.categories} selectedCategory={mockSelectedValue} />)
        expect(wrapper.find('Select').prop('defaultValue')).toEqual(mockSelectedValue)
    })

    it('should run setPriceRange if provided', () => {
        const mockPriceValue = 50
        const wrapper = shallow(<Filter setPriceRange={mockProps.setPriceRange} />)

        expect(mockProps.setPriceRange).toHaveBeenCalledTimes(0)

        wrapper.find('Slider').prop('onChange')(mockPriceValue)

        expect(mockProps.setPriceRange).toHaveBeenCalled()

        // fn returns newly selected price' value
        expect(mockProps.setPriceRange.mock.calls[0][0]).toEqual(mockPriceValue)
    })

    it('should set max price range vlaue if maxPrice provided, also should set defaultValue to currentPrice if provided', () => {
        const wrapper = shallow(<Filter maxPrice={mockProps.maxPrice} currentPrice={mockProps.currentPrice} />)
        const Slider = wrapper.find('Slider')
        expect(Slider.prop('max')).toEqual(mockProps.maxPrice)
        expect(Slider.prop('defaultValue')).toEqual(mockProps.currentPrice)
    })
})