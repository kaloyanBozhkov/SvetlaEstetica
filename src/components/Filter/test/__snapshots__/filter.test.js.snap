// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing Filter should render component 1`] = `
<Flex
  color="white"
  flexDir={
    Object {
      "base": "row",
      "xs": "column",
    }
  }
  fontSize="lg"
  fontWeight="normal"
  justifyContent="space-between"
  lineHeight="tight"
  marginTop="1rem"
>
  <Grid
    alignItems="center"
    gap="1rem"
    gridTemplateColumns="1fr 1fr"
  >
    <Text>
      Sto cercando
    </Text>
    <Select
      background="none"
      border="none"
      borderBottom="1px solid"
      borderRadius="0px"
      color="white"
      defaultValue={-1}
      focusBorderColor="none"
      fontWeight="medium"
      h="24px"
      iconSize="18px"
      lineHeight="18px"
      onChange={[Function]}
      padding="0"
      placeholder="Tutto"
      textAlign="left"
    />
  </Grid>
  <Grid
    alignItems="center"
    gap="1rem"
    gridTemplateColumns="1fr minmax(12rem, 1fr)"
    width="20rem"
  >
    <Text>
      Prezzo
    </Text>
    <Slider
      defaultValue={0}
      max={0}
      onChange={[Function]}
    >
      <SliderTrack>
        <SliderFilledTrack
          bg="white"
        />
      </SliderTrack>
      <SliderThumb />
    </Slider>
  </Grid>
</Flex>
`;
